include/zarray/zfunction.hpp

```diff
@@ -234,11 +301,111 @@ namespace xt
     }
 
     template <class F, class... CT>
-    template <std::size_t... I>
-    inline zarray_impl& zfunction<F, CT...>::assign_to_impl(std::index_sequence<I...>, zarray_impl& res, const zassign_args& args) const
+    inline zarray_impl& zfunction<F, CT...>::assign_to_impl(std::index_sequence<0>, detail::zarray_buffer_handler & buffer_handler, const zassign_args& args) const
     {
-        dispatcher_type::dispatch(detail::get_array_impl(std::get<I>(m_e), res, args)..., res, args);
-        return res;
+
+        // the input
+        auto impl_and_is_buffer = detail::get_array_impl(std::get<0>(m_e), buffer_handler, args);
+        const auto & array_impl = *(std::get<0>(impl_and_is_buffer));
+        auto is_buffer = std::get<1>(impl_and_is_buffer);
+
+
+        // the index type of the output
+        const auto result_index_type = this->get_result_type_index();
```

Comment:
I thnk we could reuse the index type of the result here. This would ensure consistency between `(a + b) + (c + d)` and `a + b + c + d` where `a` and `b` hold `float` numbers and `c` and `d` holds int numbers.

Reply:
You can forget this comment for the reasons you mentionned above.